### ISSUE TITLE
Multiple memory leaks and memory corruption

### ISSUE THREAD

**Issue by daskol** [xgfs/deepwalk-c#3]
With script and karate graph from pull request #2 I run
```bash
    convert-bcsr \
        --format mat \
        --matfile-variable-name adjmat \
        var/karate.mat var/karate.xgfs
    src/deepwalk -input var/karate.xgfs -output var/karate.emb
```
Then system sometimes dumps core. It observed the same effect if OpenMP is turned off.

**Comment by xgfs** [comment]
Are you sure this is a memory corruption and not a floating point exception? Latter should be fairly common for dimension larger than number of nodes with enough training.

**Comment by xgfs** [comment]
Reproducible with seed=0

**Comment by xgfs** [comment]
It is quite funny of a problem. There is one out-out-bounds error (you fixed it in a wrong way in #4), but ablso valgrind is angry about the perfectly fine hsm_codes initialization. Fixed both the error and the init warning.
